ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java

```diff
@@ -95,14 +115,43 @@ MessageFormatFlags getOperationFlag() {
   @Override
   void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
     //First, check if any of the existing requests have timed out.
-    cleanupExpiredInFlightRequests();
-    checkAndMaybeComplete();
-    if (!isOperationComplete()) {
-      fetchRequests(requestRegistrationCallback);
+    if (!maybeProcessDecryptCallbackAndComplete()) {
+      cleanupExpiredInFlightRequests();
+      checkAndMaybeComplete(true);
+      if (!isOperationComplete()) {
+        fetchRequests(requestRegistrationCallback);
+      }
     }
   }
 
   /**
+   * Maybe process decrypt callback and complete the operation if applicable. This is a no-op for non-encrypted blobs
+   * @return {@code true} if the operation is complete, {@code false} otherwise
+   */
+  private boolean maybeProcessDecryptCallbackAndComplete() {
+    if (onDecryptMode() && decryptCallbackResultInfo.get().resultAvailable) {
+      DecryptCallBackResultInfo resultInfo = decryptCallbackResultInfo.get();
+      if (resultInfo.exception == null) {
+        logger.trace("Successfully updating decrypt job callback results for {}", blobId);
+        operationResult = new GetBlobResultInternal(
+            new GetBlobResult(new BlobInfo(serverBlobProperties, resultInfo.result.getDecryptedUserMetadata().array()),
+                null), null);
+        state = State.Complete;
+        checkAndMaybeComplete(true);
+      } else {
+        logger.trace("Exception {} thrown on decryption for {}", resultInfo.exception, blobId);
+        setOperationException(
+            new RouterException("Exception thrown on decrypting the content for " + blobId, resultInfo.exception,
+                RouterErrorCode.UnexpectedInternalError));
+        state = State.Complete;
```

Comment:
this is kind of misleading - do you have a state for "waiting on response" or something like that ? Because the state is not really complete here - it just went back to square one and has to wait for another response.